services/src/datasets/external/netcdfcf/database.rs

```diff
@@ -0,0 +1,796 @@
+        }
+        .try_build()
+        .await?;
+
+        Ok(NetCdfCfReadAccess {
+            this: read_access,
+            provider_id,
+        })
+    }
+
+    async fn write_access(&self, provider_id: DataProviderId) -> Result<NetCdfCfWriteAccess<Tls>> {
+        let write_access = _NetCdfCfWriteAccessAsyncSendTryBuilder {
+            client: self
+                .conn_pool
+                .get_owned()
+                .await
+                .boxed_context(error::DatabaseConnection)?,
+            transaction_builder: |connection| {
+                Box::pin(async move { deferred_write_transaction(connection).await.map(Some) })
+            },
+        }
+        .try_build()
+        .await?;
+
+        Ok(NetCdfCfWriteAccess {
+            this: write_access,
+            provider_id,
+        })
+    }
+}
+
+/// Creates a read-only transaction that uses snapshot isolation
+async fn readonly_transaction<'db, 't, Tls>(
+    db_connection: &'t mut bb8_postgres::bb8::PooledConnection<'db, PostgresConnectionManager<Tls>>,
+) -> Result<tokio_postgres::Transaction<'t>>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    db_connection
+        .build_transaction()
+        .read_only(true)
+        .deferrable(true) // get snapshot isolation
+        .start()
+        .await
+        .boxed_context(error::DatabaseTransaction)
+}
+
+/// Creates a write transaction that defers constraints
+async fn deferred_write_transaction<'db, 't, Tls>(
+    db_connection: &'t mut bb8_postgres::bb8::PooledConnection<'db, PostgresConnectionManager<Tls>>,
+) -> Result<tokio_postgres::Transaction<'t>>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    let transaction = db_connection
+        .transaction()
+        .await
+        .boxed_context(error::DatabaseTransaction)?;
+
+    // check constraints at the end to speed up insertions
+    transaction
+        .batch_execute("SET CONSTRAINTS ALL DEFERRED")
+        .await
+        .boxed_context(error::UnexpectedExecution)?;
+
+    Ok(transaction)
+}
+
+#[self_referencing]
+struct _NetCdfCfReadAccess<Tls>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    client: PooledConnection<'static, PostgresConnectionManager<Tls>>,
+    #[borrows(mut client)]
+    #[covariant] // TODO: check
```

Comment:
I don't know either.

Reply:
It is, I have to remove it. The crate authors say they cannot determine it, but it would not compile if you specify it wrong.